File: src/keycap.c
#include "keycap.h"

#include "color.h"
#include "pango.h"
#include "theme.h"

#include <stdio.h>
#include <stdlib.h>

struct keycap_style {
    int padding_x;
    int padding_y;
    int gap;
    int radius;
    int border_width;
    int icon_size;
    uint32_t normal_bg;
    uint32_t normal_fg;
    uint32_t special_bg;
    uint32_t special_fg;
    uint32_t border;
};

static void rounded_rectangle(cairo_t *cairo, double x, double y, double w,
                              double h, double r) {
    if (r > w / 2.0) {
        r = w / 2.0;
    }
    if (r > h / 2.0) {
        r = h / 2.0;
    }

    cairo_new_sub_path(cairo);
    cairo_arc(cairo, x + w - r, y + r, r, -1.5707963267948966, 0.0);
    cairo_arc(cairo, x + w - r, y + h - r, r, 0.0, 1.5707963267948966);
    cairo_arc(cairo, x + r, y + h - r, r, 1.5707963267948966,
              3.141592653589793);
    cairo_arc(cairo, x + r, y + r, r, 3.141592653589793,
              4.71238898038469);
    cairo_close_path(cairo);
}

static void draw_cairo_keycap(cairo_t *cairo, const struct keycap_layout *layout,
                              const struct keycap_style *style, int radius,
                              int border_width) {
    rounded_rectangle(cairo, layout->x, layout->y, layout->width, layout->height,
                      radius);
    wsk_cairo_set_source_u32(cairo, layout->special
                                        ? style->special_bg
                                        : style->normal_bg);
    cairo_fill_preserve(cairo);

    if (border_width > 0) {
        cairo_set_line_width(cairo, border_width);
        wsk_cairo_set_source_u32(cairo, style->border);
        cairo_stroke(cairo);
    } else {
        cairo_new_path(cairo);
    }
}

static bool draw_svg_keycap(cairo_t *cairo, RsvgHandle *svg,
                            const struct keycap_layout *layout) {
    return wsk_svg_draw_to_rect(cairo, svg, layout->x, layout->y, layout->width,
                                layout->height, "key");
}

static bool draw_svg_icon(cairo_t *cairo, RsvgHandle *svg,
                          const struct keycap_layout *layout,
                          int icon_size) {
    return wsk_svg_draw_to_rect(cairo, svg, layout->icon_x, layout->icon_y,
                                icon_size, icon_size, "icon");
}

size_t wsk_measure_keycaps(cairo_t *cairo, const struct wsk_keypress *keys,
                           const struct wsk_config *config,
                           struct wsk_theme *theme, int scale,
                           uint32_t *width, uint32_t *height,
                           struct keycap_layout **out_layouts) {
    *width = 0;
    *height = 0;
    *out_layouts = NULL;

    size_t key_count = 0;
    for (const struct wsk_keypress *key = keys; key; key = key->next) {
        ++key_count;
    }
    if (key_count == 0) {
        return 0;
    }

    struct keycap_layout *layouts = calloc(key_count, sizeof(*layouts));
    if (!layouts) {
        return 0;
    }

    const struct keycap_style style = {
        .padding_x = 12,
        .padding_y = 6,
        .gap = 6,
        .radius = 8,
        .border_width = 1,
        .icon_size = 32,
    };

    const int padding_x = style.padding_x * scale;
    const int padding_y = style.padding_y * scale;
    const int gap = style.gap * scale;
    const int icon_size = style.icon_size * scale;

    int text_min_width = 0;
    int text_min_height = 0;
    int text_min_baseline = 0;
    get_text_size(cairo, config->font, &text_min_width, &text_min_height,
                  &text_min_baseline, scale, "M");

    const int min_content_width =
            icon_size > text_min_width ? icon_size : text_min_width;
    const int min_content_height =
            icon_size > text_min_height ? icon_size : text_min_height;

    size_t i = 0;
    int max_width = 0;
    int max_height = 0;
    for (const struct wsk_keypress *key = keys; key; key = key->next) {
        struct keycap_layout *layout = &layouts[i++];
        layout->key = key;
        layout->special = key->utf8[0] == '\0';
        layout->label = layout->special ? key->name : key->utf8;
        layout->icon_name = layout->special ? wsk_special_icon_name(key->name) : NULL;
        layout->icon_svg = wsk_icon_cache_get(&theme->icons, theme->base_dir,
                                              layout->icon_name);
        get_text_size(cairo, config->font, &layout->text_width,
                      &layout->text_height, &layout->text_baseline, scale, "%s",
                      layout->label);
        int content_width = layout->icon_svg ? icon_size : layout->text_width;
        int content_height = layout->icon_svg ? icon_size : layout->text_height;
        if (content_width < min_content_width) {
            content_width = min_content_width;
        }
        if (content_height < min_content_height) {
            content_height = min_content_height;
        }
        layout->width = content_width + padding_x * 2;
        layout->height = content_height + padding_y * 2;
        if (max_width < layout->width) {
            max_width = layout->width;
        }
        if (max_height < layout->height) {
            max_height = layout->height;
        }
    }

    int popup_height = max_height;
    for (i = 0; i < key_count; ++i) {
        struct keycap_layout *layout = &layouts[i];
        if ((layout->icon_svg || !layout->special) && layout->width > popup_height) {
            popup_height = layout->width;
        }
    }

    size_t total_width = 0;
    for (i = 0; i < key_count; ++i) {
        struct keycap_layout *layout = &layouts[i];
        layout->height = popup_height;
        if (layout->icon_svg || !layout->special) {
            layout->width = popup_height;
        } else {
            layout->width = max_width;
        }
        total_width += (size_t) layout->width;
    }

    *width = (uint32_t)(total_width + (key_count - 1) * (size_t) gap);
    *height = popup_height;
    *out_layouts = layouts;
    return key_count;
}

void wsk_render_keycaps(cairo_t *cairo, struct keycap_layout *layouts,
                        size_t key_count, const struct wsk_config *config,
                        struct wsk_theme *theme, int scale) {
    const struct keycap_style style = {
        .padding_x = 12,
        .padding_y = 6,
        .gap = 6,
        .radius = 8,
        .border_width = 1,
        .icon_size = 32,
        .normal_bg = 0x222222CC,
        .normal_fg = config->foreground,
        .special_bg = 0x444444CC,
        .special_fg = config->specialfg,
        .border = 0xFFFFFF33,
    };

    const int gap = style.gap * scale;
    const int radius = style.radius * scale;
    const int border_width = style.border_width * scale;
    const int icon_size = style.icon_size * scale;

    int x = 0;
    for (size_t i = 0; i < key_count; ++i) {
        struct keycap_layout *layout = &layouts[i];
        layout->x = x;
        layout->y = 0;
        x += layout->width + gap;

        if (!theme->key_svg || theme->key_svg_failed ||
            !draw_svg_keycap(cairo, theme->key_svg, layout)) {
            if (theme->key_svg && !theme->key_svg_failed) {
                fprintf(stderr, "Falling back to Cairo keycap background\n");
                theme->key_svg_failed = true;
            }
            draw_cairo_keycap(cairo, layout, &style, radius, border_width);
        }

        if (layout->icon_svg) {
            layout->icon_x = layout->x + (layout->width - icon_size) / 2;
            layout->icon_y = layout->y + (layout->height - icon_size) / 2;
            draw_svg_icon(cairo, layout->icon_svg, layout, icon_size);
            continue;
        }

        layout->text_x = layout->x + (layout->width - layout->text_width) / 2;
        layout->text_y = layout->y + (layout->height - layout->text_height) / 2;
        wsk_cairo_set_source_u32(cairo,
                                 layout->special ? style.special_fg : style.normal_fg);
        cairo_move_to(cairo, layout->text_x, layout->text_y);
        pango_printf(cairo, config->font, scale, "%s", layout->label);
    }
}

void wsk_render_keycaps_to_cairo(cairo_t *cairo, const struct wsk_keypress *keys,
                                 const struct wsk_config *config,
                                 struct wsk_theme *theme, int scale,
                                 uint32_t *width, uint32_t *height) {
    cairo_set_operator(cairo, CAIRO_OPERATOR_SOURCE);
    wsk_cairo_set_source_u32(cairo, config->background);
    cairo_paint(cairo);
    cairo_set_operator(cairo, CAIRO_OPERATOR_OVER);

    struct keycap_layout *layouts = NULL;
    size_t key_count = wsk_measure_keycaps(cairo, keys, config, theme, scale,
                                           width, height, &layouts);
    if (layouts) {
        wsk_render_keycaps(cairo, layouts, key_count, config, theme, scale);
        free(layouts);
    }
}

File: src/keycap.zig
const std = @import("std");
const c = @import("c");
const types = @import("types.zig");
const keys_mod = @import("keys.zig");
const color = @import("color.zig");
const thm = @import("theme.zig");
const pango = @import("pango.zig");
const icons = @import("icons.zig");

const Keypress = keys_mod.Keypress;
const KeycapLayout = types.KeycapLayout;
const Config = types.Config;
const Theme = types.Theme;

/// Arena for transient layout arrays. Reset on every `measureKeycaps` call.
var layout_arena: std.heap.ArenaAllocator = undefined;
var layout_arena_initialized = false;

fn ensureLayoutArena() void {
    if (!layout_arena_initialized) {
        layout_arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);
        layout_arena_initialized = true;
    }
}

/// Set to `true` by `renderKeycaps()` when any visible key is still
/// shifting its render_x toward a new target position.
/// Checked by the event loop to keep the frame loop running.
pub var shift_active: bool = false;

// ---------------------------------------------------------------------------
// Constants & types
// ---------------------------------------------------------------------------

const KeycapStyle = struct {
    padding_x: c_int = 0,
    padding_y: c_int = 0,
    gap: c_int = 0,
    radius: c_int = 0,
    border_width: c_int = 0,
    icon_size: c_int = 0,
    normal_bg: u32 = 0,
    normal_fg: u32 = 0,
    special_bg: u32 = 0,
    special_fg: u32 = 0,
    border: u32 = 0,
};

const default_keycap_style: KeycapStyle = .{
    .padding_x = 12,
    .padding_y = 6,
    .gap = 6,
    .radius = 8,
    .border_width = 1,
    .icon_size = 32,
};

const keycap_style_filled: KeycapStyle = .{
    .padding_x = 12,
    .padding_y = 6,
    .gap = 6,
    .radius = 8,
    .border_width = 1,
    .icon_size = 32,
    .normal_bg = 0x222222CC,
    .special_bg = 0x444444CC,
    .border = 0xFFFFFF33,
};

/// Default spacing between keycaps (in logical pixels at scale=1).
/// Exposed so render.zig can correctly size the layer surface.
pub const default_gap: c_int = 6;

const tau: f64 = 6.283185307179586;
const half_pi: f64 = 1.5707963267948966;

// ---------------------------------------------------------------------------
// Easing functions
// ---------------------------------------------------------------------------

/// Ease-out cubic: fast start, smooth deceleration.
fn easeOutCubic(t: f64) f64 {
    return 1.0 - std.math.pow(f64, 1.0 - t, 3.0);
}

/// Compute animation progress for a single keypress.
/// Returns a value in [0, 1] representing the fraction of animation
/// completed for entering keys, or 1 for visible keys.
fn animProgress(key: *const Keypress, now_ns: i64, duration_ns: i64) f64 {
    if (key.anim_state == .visible) return 1.0;
    const elapsed = now_ns - key.anim_start_ns;
    if (elapsed >= duration_ns) return 1.0;
    if (elapsed <= 0) return 0.0;
    return @as(f64, @floatFromInt(elapsed)) / @as(f64, @floatFromInt(duration_ns));
}

// ---------------------------------------------------------------------------
// Drawing helpers
// ---------------------------------------------------------------------------

fn roundedRectangle(cairo: ?*c.cairo_t, x: f64, y: f64, w: f64, h: f64, radius: f64) void {
    var r = radius;
    if (r > w / 2.0) r = w / 2.0;
    if (r > h / 2.0) r = h / 2.0;

    c.cairo_new_sub_path(cairo);
    c.cairo_arc(cairo, x + w - r, y + r, r, -half_pi, 0.0);
    c.cairo_arc(cairo, x + w - r, y + h - r, r, 0.0, half_pi);
    c.cairo_arc(cairo, x + r, y + h - r, r, half_pi, tau * 0.5);
    c.cairo_arc(cairo, x + r, y + r, r, tau * 0.75, tau);
    c.cairo_close_path(cairo);
}

fn drawCairoKeycap(cairo: ?*c.cairo_t, layout: *const KeycapLayout, style: *const KeycapStyle, radius: c_int, border_width: c_int) void {
    roundedRectangle(
        cairo,
        @floatFromInt(layout.x),
        @floatFromInt(layout.y),
        @floatFromInt(layout.width),
        @floatFromInt(layout.height),
        @floatFromInt(radius),
    );
    color.setSourceU32(cairo, if (layout.special) style.special_bg else style.normal_bg);
    c.cairo_fill_preserve(cairo);

    if (border_width > 0) {
        c.cairo_set_line_width(cairo, @floatFromInt(border_width));
        color.setSourceU32(cairo, style.border);
        c.cairo_stroke(cairo);
    } else {
        c.cairo_new_path(cairo);
    }
}

fn drawSvgKeycap(cairo: ?*c.cairo_t, svg: ?*c.RsvgHandle, layout: *const KeycapLayout) bool {
    return thm.svgDrawToRect(
        cairo,
        svg,
        @floatFromInt(layout.x),
        @floatFromInt(layout.y),
        @floatFromInt(layout.width),
        @floatFromInt(layout.height),
        "key",
    );
}

fn drawSvgIcon(cairo: ?*c.cairo_t, svg: ?*c.RsvgHandle, layout: *const KeycapLayout, icon_size: c_int) bool {
    return thm.svgDrawToRect(
        cairo,
        svg,
        @floatFromInt(layout.icon_x),
        @floatFromInt(layout.icon_y),
        @floatFromInt(icon_size),
        @floatFromInt(icon_size),
        "icon",
    );
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/// Measure keycap layouts: compute dimensions for each visible keypress
/// and return an array of keycap_layout structs.
pub fn measureKeycaps(
    cairo: ?*c.cairo_t,
    keys: ?*const Keypress,
    config: *const Config,
    theme: *const Theme,
    scale: c_int,
    width: *u32,
    height: *u32,
    out_layouts: *[]KeycapLayout,
) usize {
    width.* = 0;
    height.* = 0;
    out_layouts.* = &.{};

    var key_count: usize = 0;
    var key_iter = keys;
    while (key_iter) |key| {
        key_count += 1;
        key_iter = key.next;
    }
    if (key_count == 0) return 0;

    ensureLayoutArena();
    _ = layout_arena.reset(.retain_capacity);
    const allocator = layout_arena.allocator();
    const layouts = allocator.alloc(KeycapLayout, key_count) catch return 0;

    const style = &default_keycap_style;
    const padding_x = style.padding_x * scale;
    const padding_y = style.padding_y * scale;
    const gap_px = style.gap * scale;
    const icon_size_px = style.icon_size * scale;

    var text_min_width: c_int = 0;
    var text_min_height: c_int = 0;
    var text_min_baseline: c_int = 0;
    pango.getTextSize(cairo, config.font, &text_min_width, &text_min_height, &text_min_baseline, @floatFromInt(scale), "M");

    const min_content_width = @max(icon_size_px, text_min_width);
    const min_content_height = @max(icon_size_px, text_min_height);

    var i: usize = 0;
    var max_width: c_int = 0;
    var max_height: c_int = 0;
    key_iter = keys;
    while (key_iter) |key| {
        const layout = &layouts[i];
        i += 1;
        layout.key = key;
        layout.special = key.utf8[0] == 0;
        layout.label = if (layout.special) @as([*:0]const u8, @ptrCast(&key.name)) else @as([*:0]const u8, @ptrCast(&key.utf8));
        layout.icon_name = if (layout.special) icons.specialIconNameC(&key.name) else null;
        layout.icon_svg = icons.cacheGet(theme.base_dir, layout.icon_name);
        pango.getTextSize(cairo, config.font, &layout.text_width, &layout.text_height, &layout.text_baseline, @floatFromInt(scale), "%s", layout.label);

        var content_width = if (layout.icon_svg != null) icon_size_px else layout.text_width;
        var content_height = if (layout.icon_svg != null) icon_size_px else layout.text_height;
        if (content_width < min_content_width) content_width = min_content_width;
        if (content_height < min_content_height) content_height = min_content_height;
        layout.width = content_width + padding_x * 2;
        layout.height = content_height + padding_y * 2;
        if (max_width < layout.width) max_width = layout.width;
        if (max_height < layout.height) max_height = layout.height;

        key_iter = key.next;
    }

    var popup_height = max_height;
    for (0..key_count) |idx| {
        const layout = &layouts[idx];
        if ((layout.icon_svg != null or !layout.special) and layout.width > popup_height) {
            popup_height = layout.width;
        }
    }

    var total_width: usize = 0;
    for (0..key_count) |idx| {
        const layout = &layouts[idx];
        layout.height = popup_height;
        layout.width = if (layout.icon_svg != null or !layout.special) popup_height else max_width;
        total_width += @intCast(layout.width);
    }

    width.* = @intCast(total_width + (key_count - 1) * @as(usize, @intCast(gap_px)));
    height.* = @intCast(popup_height);
    out_layouts.* = layouts;
    return key_count;
}

/// Render measured keycap layouts onto the Cairo context.
pub fn renderKeycaps(
    cairo: ?*c.cairo_t,
    layouts: [*]KeycapLayout,
    key_count: usize,
    config: *const Config,
    theme: *const Theme,
    scale: c_int,
    surface_width: u32,
    content_width: u32,
) void {
    var keycap_render_style = keycap_style_filled;
    keycap_render_style.normal_fg = config.foreground;
    keycap_render_style.special_fg = config.specialfg;

    const gap_px = keycap_render_style.gap * scale;
    const radius_px = keycap_render_style.radius * scale;
    const border_width_px = keycap_render_style.border_width * scale;
    const icon_size_px = keycap_render_style.icon_size * scale;

    // Compute target X positions for all keycaps.
    // Phase 1: compute layout positions (anchor-aware start offset).
    var target_x: c_int = 0;
    const anchored_right = (config.anchor & 8) != 0; // ZWLR_LAYER_SURFACE_V1_ANCHOR_RIGHT = 8
    const anchored_left = (config.anchor & 4) != 0; // ZWLR_LAYER_SURFACE_V1_ANCHOR_LEFT = 4
    if (!anchored_left and surface_width > content_width) {
        target_x = if (anchored_right)
            @intCast(surface_width - content_width)
        else
            @intCast((surface_width - content_width) / 2);
    }

    // Write target_x into each layout.
    for (0..key_count) |i| {
        const layout = &layouts[i];
        layout.x = target_x;
        layout.y = 0;
        target_x += layout.width + gap_px;
    }

    // Compute current time for animation interpolation.
    var now_ts: c.struct_timespec = undefined;
    _ = c.clock_gettime(c.CLOCK_MONOTONIC, &now_ts);
    const now_ns = @as(i64, @intCast(now_ts.tv_sec)) * 1_000_000_000 + @as(i64, @intCast(now_ts.tv_nsec));
    const anim_duration_ns: i64 = @as(i64, @intCast(config.anim_duration_ms)) * 1_000_000;

    // Reset shift flag; set back to true if any visible key is moving.
    shift_active = false;

    for (0..key_count) |i| {
        const layout = &layouts[i];
        const key_ptr = @as(*Keypress, @constCast(layout.key orelse continue));
        const target = layout.x;

        // ── Shift interpolation ──────────────────────────────────────
        // Every frame, lerp the key's persistent render_x toward its
        // target layout position.  This gives a smooth leftward slide
        // when a new key is appended and pushes older keys over.
        const shift_lerp_speed: f64 = 0.25; // per-frame interpolation factor
        if (key_ptr.anim_state == .visible) {
            const dx = target - key_ptr.render_x;
            if (@abs(dx) <= 1) {
                key_ptr.render_x = target;
            } else {
                shift_active = true;
                const step = @as(c_int, @intFromFloat(@as(f64, @floatFromInt(dx)) * shift_lerp_speed));
                if (step == 0) {
                    key_ptr.render_x += if (dx > 0) 1 else -1;
                } else {
                    key_ptr.render_x += step;
                }
            }
        } else {
            // Entering keys snap directly to target (the entry
            // animation handles the visual fade-in/scale-up).
            key_ptr.render_x = target;
        }

        // Use the interpolated render_x as the drawing position.
        const draw_x = key_ptr.render_x;
        // Temporarily override layout.x so drawSvgKeycap / drawCairoKeycap
        // use the shifted position rather than the raw target.
        const layout_x_saved = layout.x;
        layout.x = draw_x;

        const progress = animProgress(key_ptr, now_ns, anim_duration_ns);
        const eased = easeOutCubic(progress);
        const is_entering = eased < 1.0;

        if (is_entering) {
            c.cairo_push_group(cairo);
        }

        c.cairo_save(cairo);
        if (is_entering) {
            const cx = @as(f64, @floatFromInt(draw_x)) + @as(f64, @floatFromInt(layout.width)) / 2.0;
            const cy = @as(f64, @floatFromInt(layout.y)) + @as(f64, @floatFromInt(layout.height));
            c.cairo_translate(cairo, cx, cy);
            c.cairo_scale(cairo, 1.0, eased);
            c.cairo_translate(cairo, -cx, -cy);
        }

        const svg_succeeded = theme.key_svg != null and !theme.key_svg_failed and drawSvgKeycap(cairo, theme.key_svg, layout);
        if (!svg_succeeded) {
            if (theme.key_svg != null and !theme.key_svg_failed) {
                std.log.err("Falling back to Cairo keycap background", .{});
                @constCast(theme).key_svg_failed = true;
            }
            drawCairoKeycap(cairo, layout, &keycap_render_style, radius_px, border_width_px);
        }

        if (layout.icon_svg != null) {
            layout.icon_x = draw_x + @divTrunc(layout.width - icon_size_px, 2);
            layout.icon_y = layout.y + @divTrunc(layout.height - icon_size_px, 2);
            _ = drawSvgIcon(cairo, layout.icon_svg, layout, icon_size_px);
        } else {
            layout.text_x = draw_x + @divTrunc(layout.width - layout.text_width, 2);
            layout.text_y = layout.y + @divTrunc(layout.height - layout.text_height, 2);
            color.setSourceU32(cairo, if (layout.special) keycap_render_style.special_fg else keycap_render_style.normal_fg);
            c.cairo_move_to(cairo, @floatFromInt(layout.text_x), @floatFromInt(layout.text_y));
            pango.printf(cairo, config.font, @floatFromInt(scale), "%s", layout.label);
        }

        c.cairo_restore(cairo);

        if (is_entering) {
            c.cairo_pop_group_to_source(cairo);
            c.cairo_paint_with_alpha(cairo, eased);
        }

        // Restore layout.x for the caller (render.zig uses it for
        // content_width calculations; leaving it as draw_x would
        // cause the reserved width to drift on every frame).
        layout.x = layout_x_saved;
    }
}

/// Render keycaps directly to a Cairo context (convenience wrapper).
pub fn renderKeycapsToCairo(
    cairo: ?*c.cairo_t,
    keys: ?*const Keypress,
    config: *const Config,
    theme_param: *const Theme,
    scale: c_int,
    width: *u32,
    height: *u32,
) void {
    c.cairo_set_operator(cairo, c.CAIRO_OPERATOR_SOURCE);
    color.setSourceU32(cairo, config.background);
    c.cairo_paint(cairo);
    c.cairo_set_operator(cairo, c.CAIRO_OPERATOR_OVER);

    var layouts: []KeycapLayout = &.{};
    const key_count = measureKeycaps(cairo, keys, config, theme_param, scale, width, height, &layouts);
    if (layouts.len > 0) {
        renderKeycaps(cairo, layouts.ptr, key_count, config, theme_param, scale, width.*, width.*);
    }
}
